unified-codes/apps/web/src/sagas/ExplorerSaga.ts

```diff
@@ -1,15 +1,31 @@
 import { call, put, takeEvery, all } from 'redux-saga/effects';
 
-import { IEntity, AlertSeverity, IAlert } from '@unified-codes/data';
+import {
+  AlertSeverity,
+  Entity,
+  IAlert,
+  IPaginatedResults,
+  IPaginationRequest,
+} from '@unified-codes/data';
 
-import { EXPLORER_ACTIONS, ExplorerActions, AlertActions, IExplorerAction } from '../actions';
+import {
+  EXPLORER_ACTIONS,
+  ExplorerActions,
+  AlertActions,
+  IExplorerAction,
+  IExplorerFetchDataAction,
+} from '../actions';
 
-const GET_ENTITIES = `
+const getEntitiesQuery = (first: number, offset?: number) => `
   {
-    entities {
-      code
-      description
-      type
+    entities(offset: ${offset} first: ${first}) {
+      data {
+        code
+        description
+        type
+      },
+      hasMore,
+      totalResults,
```

Comment:
> Could totalResults be derived at the front-end by counting the number of entities returned?

Only by storing all entities in the front-end which is the sort of inefficiency that pagination is trying to avoid!

Sure, if you have all entities then you can derive these, and don't need them really. you could provide all entities to the explorer and use local state to manage the view window. Given the size of our dataset we could happily enough do that.

I was thinking that we'd want to make use of some of dgraph's capabilities, so we'll need push querying down to the db level. That doesn't preclude keeping pagination local though.

Thoughts on fetching all data and storing in front end? Could then push pagination to the local state level and refer back to the dgraph server only when searching.

Reply:
> Only by storing all entities in the front-end which is the sort of inefficiency that pagination is trying to avoid!

Ah, so `totalResults` is the total count of all entities in the DB, or all entities that fit the search criteria? I think storing all front-end is the way were going, but as an aside, I think it might also be possible to add a query to the schema for that (would require an extra trip, but I think that would be a rather small cost versus the simplification that would come from decoupling the "metadata" out of the entity query? I think I'd need to read up a bit more on how to do pagination in `graphql`, `dgraph` to be up-to-speed and offer useful feedback here, so just another reason to go for pulling all the entities in for now!

My thinking is to keep it as simple as possible and deal with optimisations later, even if it seems obvious that an optimisation will be needed. I think YAGNI is a good principle for us here, as an unoptimised naive solution will be a good starting point for us all to understand the problem a bit more, rather than trying to go from zero to perfectly optimised. Does that make sense? Agree? 

So with that in mind, I would suggest pulling everything and filtering locally (for now). Then once searching/pagination is implemented locally (if I'm still up-to-date, data filtering is done locally too, with the assumption all entities are in the store?), we can then look at how we might optimise it (lots of options and can be done incrementally, I think).

@katherine-sussol @mark-prins Have I missed anything? Agree? Disagree?